### ISSUE TITLE
universal wheels on PyPI ?

### ISSUE THREAD

**Issue by johnyf** [dabeaz/ply#120]
`setup.cfg` [selects](https://github.com/dabeaz/ply/blob/b515c5fa1003a2a575b5c175831142da94e21ee5/setup.cfg#L1) [universal wheels](https://packaging.python.org/distributing/#universal-wheels) (introduced by #81), but only the source distribution appears [on PyPI](https://pypi.python.org/pypi/ply/3.10). Are there plans to upload to PyPI also wheel files in the future?

**Comment by dabeaz** [comment]
Didn't even cross my mind.  Is there some tangible benefit to it given that PLY is already pure Python and quite small?

**Comment by johnyf** [comment]
I have been studying about wheels only recently (for a use case with Cython extensions, so there it makes a clear difference) and noticed from [pythonwheels](http://pythonwheels.com/) that there are no wheel files for `ply` on PyPI. Looking at the repository, I saw `setup.cfg`, so I wondered what the reason could be for preferring `sdist` to wheels.

The only potential advantages I am aware of are:

1. slightly faster installation from wheel than `sdist` (`pip` caches wheels for all packages it downloads, so this applies only once), and
2. avoiding execution of `setup.py` at installation time (based on [this discussion](http://stackoverflow.com/questions/31401762/python-packaging-wheels-vs-tarball-tar-gz) that cites the [rationale in PEP 427](https://www.python.org/dev/peps/pep-0427/#rationale)).

Item 1 is a general remark--I haven't profiled `ply`'s installation. A potential disadvantage may be having to upload more than one thing to PyPI. Simple is better than complex, so only `sdist` may be a better choice.

In general, another disadvantage (**not applicable** to `ply` though, as of b791b089b7d368592838ebbb8172fecc74001816) is that the wheel can become outdated. For example, if a package depends on `ply` and that package's wheel contains parser tables, then those tables can become outdated when a later `ply` version is released (even though the package may still be compatible with that version), and they won't be rebuilt at installation time.

**Comment by gvalkov** [comment]
There is indeed a small advantage in terms of file size and installation speed:

```
52K	dist/ply-3.11-py2.py3-none-any.whl
152K	dist/ply-3.11.tar.gz
```

```
pip install --no-index -f ./dist --no-binary :all: ply	1.05s user 0.09s system 91% cpu 1.250 total
pip install --no-index -f ./dist ply			0.39s user 0.04s system 97% cpu 0.441 total
```

This would be nice to have, imho.

**Comment by dabeaz** [comment]
I don't have the personal bandwidth or interest to manage all sorts of different distribution formats for PLY.  For a pure-python library that only consists of a few files, I don't see much of a benefit.   If someone wants to a contribute a Makefile that just does it without me having to worry about it, I'll consider it.

**Comment by gvalkov** [comment]
That's understandable, especially since these things tend to change every now and then. Still, from the looks of it, wheels are the new standard (i.e. [PEP427](https://www.python.org/dev/peps/pep-0427/)). 

I've made a small makefile that combines uploading an sdist and a wheel into a single `upload` target (PR #128). Please note that you might have to do a `python -m pip install wheel` beforehand. Thanks.

**Comment by johnyf** [comment]
I will close this issue, because I couldn't find any advantage to distributing wheels for a pure Python package. For example, `networkx` is present only as source on PyPI.

**Comment by johnyf** [comment]
I now have learned why wheel files would better be used only when unavoidable: the presence of both source and wheel distributions creates two possible configurations for `pip install` and `pip download`. More configurations mean more ways things can go wrong, and things *do* go wrong. The number of configurations grows exponentially with the number of dependencies that offer both source and wheel distributions.

Why does this matter? Testing installation in different environments is a good idea. That a dependency offers some wheels is insufficient an argument to claim that one need not test with the dependency installed from source. The reason is that the collection of wheel files available from PyPI likely covers only some Python versions, and only some operating systems. In all other environments, `pip` will fetch a source tarball and attempt to install from it (this is very likely when testing multiple Python versions on Travis CI). So one does have to test with dependencies installed from both their source and their wheel distributions.

This situation increases the complexity of the testing configuration (e.g., of `.travis.yml`), and is fragile. For example, a script that does a `pip download --no-dependencies foo` in order to obtain the source of a dependency, and then unpacks the source and builds `foo` with its C extensions, will fail as soon as package `foo` starts offering wheel files on PyPI (assuming those wheels match the installation environment). The failure is caused because `pip` will now download a wheel, instead of a source tarball, and then fail to untar, because no `tar.gz` is present. Of course, updating the script to `pip download --no-dependencies foo --no-binary foo` avoids this issue, but the point is that even (seemingly) very simple things become complicated with wheels.

So, unless a package definitely benefits from distributing wheels, due to having C extensions, there are disadvantages in distributing wheels. Simple is better than complex [PEP 20].
